advanced/profiling.md

```diff
@@ -0,0 +1,254 @@
+# CPU and Memory Profiling
+Code profiling is the process of analyzing a program to understand its runtime behavior using performance characteristics like CPU/GPU usage, memory usage, IO operations, and total runtime. Profilers can help developers identify bottlenecks such as high CPU usage, memory usage, and runtime. There are profilers for most languages and environments, but we will focus on Python profiling tools in this guide.
+
+
+## Sampling vs. Deterministic profilers
+There are two main types of profilers and you should be aware of the differences between them. For small programs, you may not notice a difference, but for larger programs, the choice of profiler can have a significant impact on the performance of your program.
+- Deterministic (event-based)
+    - Can have high overhead
+    - Trace every function call
+    - Python ex: `cprofile`, `time`, `line_profiler`, `memray`
+- Sampling (statistical)
+    - Sample the stack periodically
+    - Non deterministic
+    - Add less overhead
+    - Python ex: `austin`, `pyinstrument`, `py-spy`
+
+## Python and iPython/Jupyter profilers
+Let's test out these tools with a sample program (`random_primes_sum.py`) that is CPU and memory intensive:
+```python
+import random
+import math
+
+def random_primes_sum():
+    random_list = [random.uniform(1, 10000) for _ in range(1000000)]
+    sorted_list = sorted(random_list)
+    transformed_list = [math.sqrt(x) + math.log(x) for x in sorted_list]
+    def is_prime(n):
+        if n <= 1:
+            return False
+        for i in range(2, int(math.sqrt(n)) + 1):
+            if n % i == 0:
+                return False
+        return True
+    return sum(1 for x in transformed_list if is_prime(int(x)))
+
+random_primes_sum()
+```
+
+### time
+- Simple single run timer
+- iPython magics `%time` and `%%time`
+- Useful for long running operations
+- Minimal overhead
+- Output can vary due to current system load and other factors (see timeit for accurate averaged timing)
+
+
+```python
+In [1]: %time random_primes_sum()
+Out[1]: CPU times: user 521 ms, sys: 26.2 ms, total: 547 ms
+        Wall time: 560 ms
+```
+
+### timeit
+- Multiple run timer
+- Useful when operation is fast and you want to average over several runs for accurate timing
+- iPython magics `%timeit` and `%%timeit`
+- Configurable flags such as #iterations, precision, and saving output to variable (see the [ipython docs for all flags](https://ipython.readthedocs.io/en/stable/interactive/magics.html#magic-timeit))
+- Can also be used on python scripts (see [python docs for examples](https://docs.python.org/3/library/timeit.html))
+- Warning: Watch out for operations that modify a global state such as `file.read()` or `list.sort()`. E.g. `%timeit arr.sort()` will display an inaccurate timing since the list will be sorted in-place on the first iteration, then subsequent iterations will be near instant.
```

Comment:
````suggestion

```{warning}
Watch out for operations that modify a global state such as `file.read()` or `list.sort()`. E.g. `%timeit arr.sort()` will display an inaccurate timing since the list will be sorted in-place on the first iteration, then subsequent iterations will be near instant.
```

````

Just a small tweak to make this more of a callout using admonition with myst markdown syntax: https://myst-parser.readthedocs.io/en/latest/syntax/admonitions.html